### PR TITLE
Added support to TS and Footer on error

### PR DESCRIPTION
**Pull request by extmkv**
### Checklist
- [x] I've read and followed the [Contribution Guidelines](https://github.com/bitrise-steplib/.github/blob/main/CONTRIBUTING.md)
- [x] `step.yml` and `README.md` is updated with the changes (if needed)

### Version
Requires a *MINOR* [version update](https://semver.org/)

### Context

Adds support to update messages using slack step.
Also adds support to the footer message and icon on error.

### Changes
Added 3 new properties:

`ts`: When ts is provided and when using `api_token` it will update the original message with new content. 
To understand how the slack update method works, please check [here](https://api.slack.com/methods/chat.update).

` footer_on_error`: Works in the same way as the others on error.
`footer_icon_on_error`: Works in the same way as the others on error.

### Investigation details

N/A

### CODE REVIEW

**Comment by extmkv** [comment]
@godrei can you have a look?

**Comment by extmkv** [comment]
@hisaac can you have a look?

**Comment by hisaac** [comment]
Hi @extmkv, I don't work on this anymore so I won't be able to help. Perhaps @DamienBitrise could help, or recommend someone who could review?

**Comment by RamithaW** [comment]
@DamienBitrise is this something you can help review? We would love to have have this feature on the bitrise-steplib.

File: main.go [hunk 1]
```diff
@@ -35,36 +35,41 @@ type Input struct {
...
+	Message           string `env:"message"`
+	MessageOnError    string `env:"message_on_error"`
+	ImageURL          string `env:"image_url"`
+	ImageURLOnError   string `env:"image_url_on_error"`
+	ThumbURL          string `env:"thumb_url"`
+	ThumbURLOnError   string `env:"thumb_url_on_error"`
+	Footer            string `env:"footer"`
+	FooterIcon        string `env:"footer_icon"`
+	FooterOnError     string `env:"footer_on_error"`
+	FooterIconOnError string `env:"footer_icon_on_error"`
```
Comment (godrei) [inline main.go:61]: I suggest following the existing pattern, that the `OnError` version of the input is right after its regular version.

```suggestion
	Footer            string `env:"footer"`
	FooterOnError     string `env:"footer_on_error"`
	FooterIcon        string `env:"footer_icon"`
	FooterIconOnError string `env:"footer_icon_on_error"`
```

File: main.go [hunk 2]
```diff
@@ -35,36 +35,41 @@ type Input struct {
...
+	Fields            string `env:"fields"`
+	Buttons           string `env:"buttons"`
 
 	// Status
 	BuildStatus         string `env:"build_status"`
 	PipelineBuildStatus string `env:"pipeline_build_status"`
 
 	// Step Outputs
 	ThreadTsOutputVariableName string `env:"output_thread_ts"`
+	TsOutputVariableName string `env:"output_ts"`
```
Comment (godrei) [inline main.go:72]: Do you think this input is needed? For me, it seems that it is unused currently.
The `ThreadTsOutputVariableName` seems to be a similar input and this one is used to export a Step Output variable specified by the input: https://github.com/extmkv/steps-slack-message/blob/master/outputs.go#L41

File: step.yml [hunk 3]
```diff
@@ -315,6 +326,21 @@ inputs:
       description: |
         Renders a small icon beside the footer text
         It will be scaled down to 16px by 16px.
+  - footer_on_error: "Bitrise"
```
Comment (godrei) [inline step.yml:329]: NOTE: I suggest following the existing pattern, that the `OnError` version of the input is right after its regular version.

**Review by godrei — COMMENTED** [review]
Hi @extmkv, sorry for the late response.
I reviewed the code changes and have a few comments, please check it out.

File: step.yml [hunk 4]
```diff
@@ -186,6 +186,17 @@ inputs:
...
+
+  - ts:
+    opts:
+      title: Message Timestamp
+      description: Updates the message.
+  - ts_on_error:
+    opts:
+      title: Message Timestamp if the build failed
+      description: Updates the message if the build failed.
+      category: If Build Failed
```
Comment (godrei) [inline step.yml:198]: I would also suggest a bit more descriptive step input properties.

Screenshot 2023-11-17 at 15 46 26


```suggestion
  - ts:
    opts:
      title: Message Timestamp
      summary: Timestamp of the message to be updated.
      description: |-
        Timestamp of the message to be updated.
        
        When **Message Timestamp** is provided an existing Slack message will be updated, identified by the provided timestamp.  
        Example: `"1405894322.002768"`.
  - ts_on_error:
    opts:
      title: Message Timestamp if the build failed
      summary: Timestamp of the message to be updated if the build failed.
      description: |-
        Timestamp of the message to be updated if the build failed.
        
        When **Message Timestamp if the build failed** is provided an existing Slack message will be updated, identified by the provided timestamp.  
        Example: `"1405894322.002768"`.
      category: If Build Failed
```

File: main.go (hunk 1 above)
Comment (extmkv) [reply main.go]: Fixed

File: step.yml (hunk 4 above)
Comment (extmkv) [reply step.yml]: Fixed

File: main.go (hunk 2 above)
Comment (extmkv) [reply main.go]: You're completely right, I was running a few tests and forgot to remove this.

**Comment by extmkv** [comment]
@godrei I've fixed your comment.

**Comment by godrei** [comment]
Hi @extmkv , thanks for the changes.
I did some manual testing, to try out the new feature and I faced an interesting difference between the `chat.postMessage` and `chat.update` endpoints.

I used this workflow to post a message to a channel and then update the original message:
```
  slack:
    steps:
    - git::https://github.com/extmkv/steps-slack-message.git@master:
        title: Send message to be updated
        is_skippable: false
        inputs:
        - api_token: "$SLACK_API_TOKEN"
        - channel: "$SLACK_CHANNEL"
        - from_username: step-dev-test
        - text: Testing slack message update
        - output_thread_ts: THREAD_TS
        - is_debug_mode: 'yes'
    - git::https://github.com/extmkv/steps-slack-message.git@master:
        title: Update message
        is_skippable: false
        inputs:
        - api_token: "$SLACK_API_TOKEN"
        - channel: $SLACK_CHANNEL
        - from_username: step-dev-test
        - text: Slack message updated
        - ts: "$THREAD_TS"
```

The messages are connected through the `THREAD_TS` output of the first step. I expected the original message's text to change from `Testing slack message update` to `Slack message updated`.

What I experienced is that the first message was posted, but it wasn't updated, even if both steps were successful. 
After adding some debug logging around the HTTP calls, it turned out that the client returned status OK for the update message with the following body:
```
{"ok":false,"error":"channel_not_found"}
```

And here was the source of the issue: I used a channel's name as the `channel` input for the step and it was working fine for posting a message, but the update endpoint requires a channel ID.

After updating the channel step input, I managed to verify the new functionality, so we can merge and release the changes.
I will try to allocate some time to improve how the message posts and updates are verified.

**Review by godrei — APPROVED** [review]

**Pull request merged**